### PR TITLE
docs/enclave-tls: add design doc

### PR DESCRIPTION
**Pull request by haosanzi**
Fixes: #748
Signed-off-by: Shirong Hao <shirong@linux.alibaba.com>

### CODE REVIEW

File: docs/design/enclave-tls/design-zh.md [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+- 能够支持不同类型HW-TEE间的双向TLS认证。
+
+# 架构设计
+
+![architecture.png](architecture.png)
+
+Enclave-TLS的架构如上图所示，Enclave TLS 提供了五个API给上层的应用和服务使用，五个API能够实现可信安全信道的建立和数据的传输。Enclave TLS API的实现依赖于核心层与四类实例的实现。
+
+为了提供更好的安全性，核心层与四类实例插件都运行在HW TEE环境中。由于Enclave-TLS架构需要支持和灵活选择多种可能性（例如：TLS库，加密库，Enclave形态等），因此根据功能逻辑区分出不同的实例插件是必不可少的。下面简单介绍一下核心层和实例插件的作用。
+- **核心层**。负责整体的控制作用，控制数据流的流向。
```
Comment (zhiminghufighting) [inline docs/design/enclave-tls/design-zh.md:52]: **核心层**。这里跟“：”是否更合适？

File: docs/design/enclave-tls/design-zh.md [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+- **TLS Wrapper实例**。负责完成真正的TLS session管理和网络传输。
+- **Attester 实例**。从本地平台运行环境中收集证明材料，通常要有Enclave配合生成quote数据。
+- **Verifier 实例**。负责验证收到的各种格式的quote数据，可能收到来自另一个不同的机密计算硬件平台发来的quote。也可能存在verifier与attester必须是相同平台的情况，比如verifier使用SGX ECDSA QVE的情况：强制要求当前运行环境必须支持SGX ECDSA，否则无法启动QVE来验证SGX ECDSA quote。
```
Comment (zhiminghufighting) [inline docs/design/enclave-tls/design-zh.md:55]: The following three lines have the same question?

File: docs/design/enclave-tls/design-zh.md [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+传统上扩展证书需要由 CA 对其进行签名。 但是，由于我们建议使用硬件TEE 作为信任根，我们可以简单地对证书进行自签名。 我们不依靠 CA 将域名绑定到服务器的身份/密钥，而是依靠硬件平台来提供身份。
+
+嵌入到 Enclave-TLS 证书中的证明材料根据Attester 实例的不同有所不同。具体信息需要参考不同的enclave实例的远程报告信息。
+
+3. 客户端验证证书
+
+Enclave-TLS 证书的接收者必须验证Evidence来确定特定平台上Enclave的可信度。 具体步骤因认证方法而异。 在协议方面，验证发生在标准 TLS 握手期间。 如果任何验证步骤失败，端点将中止握手以终止 TLS 连接。除此之外，验证者必须确保Enclave-TLS证书的公钥的hash确实存在于enclave的报告中，然后验证者将从证书中获取到的enclave的身份信息和验证者已有的基准信息进行比较。
+
+## 异构HW-TEE节点的安全可信通信
+
```
Comment (zhiminghufighting) [inline docs/design/enclave-tls/design-zh.md:92]: If this function is not implemented or it is a future plan, please specify out it.

**Review by YangLiang3 — APPROVED** [review]

File: docs/design/enclave-tls/design-zh.md [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+在某些情况下，客户端和服务端可以运行在不同的机密计算平台上，可能存在服务端获取了不同的quote类型的认证报告。如何有效的对多种类型的quote实例进行验证并建立安全信道是一个问题。
+
+为了解决这个问题，Enclave-TLS将验证不同quote类型的逻辑抽象成对应Verifier实例，通过对象标识符（Object Identifier，OID）来区分接收到的证书类型，进而选择不同的Verifier实例进行证书的验证（异构HW-TEE节点的安全可信通信正在实现中）。
```
Comment (fenghao2021) [inline docs/design/enclave-tls/design-zh.md:95]: 建议增加OID的定义说明

File: docs/design/enclave-tls/design.md [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+
+According to the respective responsibilities and constraints of each instance, we have designed common metadata and APIs for each type of instance.
+
+## TLS Wrapper instance
+
+Each TLS Wrapper instance is designed with the following metadata and API to complete the functions of TLS session management and network transmission. The following is the meaning of each metadata and method.
+
+```c
+typedef struct {
+    uint8_t api_version;
```
Comment (liangzhou121) [inline docs/design/enclave-tls/design.md:213]: I think the format here need to align.

File: docs/design/enclave-tls/design.md [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+
+`libtls_wrapper_<type>.so` will register metadata and methods to the Enclave-TLS core when it is initialized.
+
+```c
+tls_wrapper_err_t tls_wrapper_register(const tls_wrapper_opts_t *opts);
+```
+
+## Enclave Attester instance
+
+Each Attester instance has designed the following metadata and API to be responsible for collecting evidence from the local platform operating environment. The following is the meaning of each metadata and method.
```
Comment (liangzhou121) [inline docs/design/enclave-tls/design.md:246]: change Attester instance has designed the -> Attester instance is designed with ?
To align with follow "Enclave Verifier instance" description.

File: docs/design/enclave-tls/design.md [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+
+`libverifier_<type>.so` will register metadata and methods to the Enclave-TLS core when it is initialized.
+
+```c
+enclave_verifier_err_t enclave_verifier_register(const enclave_verifier_opts_t *opts);
+```
+
+## Crypto Wrapper instance
+
+Each Crypto Wrapper instance designs the following metadata and API to cooperate with other instances to complete operations related to cryptographic algorithms. The following is the meaning of each metadata and method.
```
Comment (liangzhou121) [inline docs/design/enclave-tls/design.md:321]: Each Crypto Wrapper instance designs the following 
The three instances' description here maybe need to align.

**Review by zhiminghufighting — APPROVED** [review]

**Review by fenghao2021 — APPROVED** [review]

**Review by liangzhou121 — APPROVED** [review]

**Pull request merged**